### PR TITLE
Artillery performance testing

### PR DESCRIPTION
**Pull request by s373nZ**
This is an initial version of a small performance testing suite using [Artillery.io](https://www.artillery.io). It's currently tuned toward a `t3.small` instance (changed from production's `t2.small` as it has more CPU and is same cost) and features some general HTTP traffic simulation as well as a REPL capacity testing profile which performs actual cryptographic operations over a delayed wait time to simulate actual CPU pressure.

In practice, I've been using the `make test-perf-load` profile to run tests against the Autoscaling changes. The tests are probably waaay tooo aggressive in terms of representing actual user traffic, and it's possible the stress test could be removed. Currently, even the load test overwhelms the `t3.small` instance for about 2 minutes until fast-scaling emergency backups come online.

The test also highlights a number of other performance issues (on my test setup) that could be investigated, database activity or connection limits, for example.

Tests aren't perfect. They could be tuned to represent more realistic user traffic, with more accurate wait times (`think` actions), or traffic pattern that is more representative of actual users. It would be cool to increase the CloudWatch log retention setting during a class, or the next BOSS program, and have take a shot at creating test scenarios based on analysis of that traffic...

Either way, the current test fulfills the purpose of slamming the `staging` setup to provoke autoscaling to prove it works. Taking this out of WIP, at least.

### CODE REVIEW

File: test/performance/profiles/http-tests.yaml [hunk 1]
```diff
@@ -0,0 +1,466 @@
+# =============================================================================
+# CONSOLIDATED HTTP API TESTS
+# =============================================================================
+# This file contains all HTTP-based performance test profiles using Artillery's
+# environments feature. Each environment (smoke/load/stress) has its own phases
+# and SLAs, but shares the same scenario definitions.
+#
+# Usage:
+#   artillery run -e smoke /scripts/profiles/http-tests.yaml
```
Comment (satsie) [inline test/performance/profiles/http-tests.yaml]: Is this the correct path to the file? I don't see a top level `scripts` directly but maybe that is configured in the CI

File: Makefile [hunk 2]
```diff
@@ -21,7 +21,31 @@ stop-deps:
...
+# Artillery Performance Tests
+#
+# Start a performance testing server on 8001 beforehand with:
+#   `yarn test:performance`
+#
+# TODO: After upgrading NodeJS to a more recent version,
+#   consider installing Artillery via NPM and running
+#   via script.
+ARTILLERY_TARGET ?= http://localhost:8001
+ARTILLERY_WS_TARGET ?= ws://localhost:8001
```
Comment (jrakibi) [inline Makefile]: on macOS this default fails when running `make test-perf-smoke`, so I have to change `localhost` to `host.docker.internal` to make it work

File: test/performance/README.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+| `REPL_DIST_LONG` | `0.08` | Probability of 15-29s execution (heavy computation) |
+| `REPL_DIST_TIMEOUT` | `0.02` | Probability of 31-40s execution (triggers 30s timeout) |
+
+```bash
+# Run with higher timeout rate for stress testing
+REPL_DIST_TIMEOUT=0.10 REPL_DIST_SHORT=0.62 make test-perf-stress
+
+# Run with no timeouts for quick validation
+REPL_DIST_TIMEOUT=0 make test-perf-smoke
+```
```
Comment (jrakibi) [inline test/performance/README.md:88]: are these `REPL_*` vars supposed to be in the makefile ? cause I can't see them forwarded in `docker run`, so I'm not sure if the container will actually read them

think we need to add 
```
-e REPL_DIST_SHORT 
-e REPL_DIST_MEDIUM 
-e REPL_DIST_LONG 
-e REPL_DIST_TIMEOUT 
```
in the makefile

File: test/performance/README.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+## Quick Start
+
+```bash
+# Start dependencies and server
+make start-deps
+yarn dev
+
+# Run smoke test (quick validation)
+make test-perf-smoke
```
Comment (jrakibi) [inline test/performance/README.md:20]: tried following this guide, `yarn dev` runs the server at port 8000 but the makefile targets 8001

I get an `ECONNREFUSED` error when running `yarn dev`, should we change it to `yarn build && yarn test:performance` ?
```suggestion
# Start dependencies and server
make start-deps
yarn build && yarn test:performance

# Run smoke test (quick validation)
make test-perf-smoke
```

File: Makefile (hunk 2 above)
Comment (s373nZ) [reply Makefile]: Wasn't aware of this difference. Changed it to `127.0.0.1`. Works on Linux, hopefully it works on OSX as well?

File: test/performance/README.md (hunk 3 above)
Comment (s373nZ) [reply test/performance/README.md:88]: Nice find. Updated the `Makefile` to pass these vars through.

File: test/performance/README.md (hunk 4 above)
Comment (s373nZ) [reply test/performance/README.md:20]: Yeah, this got missed. Updated the README, thanks.

**Comment by s373nZ** [comment]
@jrakibi Thanks so much for taking a look, despite running into some quick operational issues! This work can definitely benefit from your attention if you have any review or testing time to spare.

More generally, I'm curious if you have any initial feedback on the following:
- I'm not super confident the testing profiles are tuned very accurately. If they *are*, I'm thinking about removing the `stress` test as it's waaaay to aggressive - the load test breaks the autoscaling for `t3.small` easily. No sense in maintaining the `stress` test if it's too wild to be useful.
- If you or anyone else has feedback on the testing "scenarios" or traffic patterns, it would be nice to have them match reality. Suggestions super welcome on this part, although learning `artillery` tooling is involved a little.

**Comment by jrakibi** [comment]
> I'm thinking about removing the stress test as it's waaaay to aggressive

okay that makes sense, if load test already breaks t3.small, don't think the stress test will tell us any new info.

> If you or anyone else has feedback on the testing "scenarios" or traffic patterns

about traffic patterns, even tho I've never run a Sav Sat workshop myself, I think one scenario that could happen is many users being in sync on what they're doing on the platform. so instead of them doing different things, 90% of them might be doing REPL stuff at the same time.

File: test/performance/processors/functions.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    console.log(`[${context.vars.testId || 'unknown'}] Status: ${response.statusCode}`);
+  }
+  return next();
+}
+
+// =============================================================================
+// REPL Code Generation with Configurable Timing Distribution
+// =============================================================================
+//
+// Environment variables to control execution time distribution:
```
Comment (satsie) [inline test/performance/processors/functions.js:53]: What happens if the probability of these `REPL_DIST_*` env vars doesn't add up to 1?

File: test/performance/profiles/repl-capacity.yaml [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+          - send:
+              action: "repl"
+              payload:
+                # console.log(Date.now());
+                code: "Y29uc29sZS5sb2coRGF0ZS5ub3coKSk7"
+                language: "javascript"
+          - think: 3
+        count: 3
+
+  - name: "JavaScript REPL - Timeout (30s+)"
```
Comment (satsie) [inline test/performance/profiles/repl-capacity.yaml]: Really happy to see automated testing for this. Thanks for including this edge case!

File: test/performance/processors/functions.js [hunk 7]
```diff
@@ -91,25 +91,55 @@ function getWeightedDelay() {
 
 /**
  * Generate JavaScript REPL code with variable execution time.
+ * Runs secp256k1 point multiplications (G.mul) for the target duration, mirroring
+ * the actual CPU work saving-satoshi users perform in production.
```
Comment (satsie) [inline test/performance/processors/functions.js:101]: Nice!

File: test/performance/profiles/repl-capacity.yaml [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+# Differences from other profiles:
+#   - Uses WebSocket engine instead of HTTP
+#   - Tests container orchestration, not API endpoints
+#   - Critical for AWS instance sizing decisions
+#   - Longer phases to observe CPU credit depletion
+#
+# Capacity estimates for t2.small (2GB RAM, 20% CPU baseline):
+#   - Burst:      ~3-4 concurrent containers (with credits)
+#   - Sustained:  ~2 concurrent containers (after credit depletion)
+#   - Critical:   >5 concurrent containers (memory exhaustion)
```
Comment (satsie) [inline test/performance/profiles/repl-capacity.yaml:54]: I'm highlighting this as a note to self. In October 2024 we ran a 50 person Saving Satoshi workshop at TABConf. By the halfway point, the backend had become wildly unstable and only seemed to work for people 20% of the time. I believe at the time we were on an even smaller instance (t2.micro?). We had always suspected that the resources were exhausted but never knew 1. the max number of concurrent users that the system could handle (which would allow us to set attendee limits on workshops and provide some idea for  max number of concurrent users if we sized up the instance), and 2. if the constraint was with the CPU or memory. 

Having these numbers solves a huge mystery, and makes me A LOT more confident in our ability to do future workshops. I'll be passing this along to the folks that helped with the 2024 workshop. Can't believe we finally have insight into this. You don't know how much this means and how much peace of mind this provides. Thank you!!

File: test/performance/profiles/repl-capacity.yaml [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+          - send:
+              action: "repl"
+              payload:
+                # console.log(Date.now());
+                code: "Y29uc29sZS5sb2coRGF0ZS5ub3coKSk7"
+                language: "javascript"
+          - think: 3
+        count: 3
+
+  - name: "JavaScript REPL - Timeout (30s+)"
```
Comment (satsie) [inline test/performance/profiles/repl-capacity.yaml:159]: Really happy to see automated testing for this because it has come up multiple times as something we think works, but have often retested manually. Thanks for including this edge case!

**Comment by satsie** [comment]
> I'm thinking about removing the stress test as it's waaaay to aggressive - the load test breaks the autoscaling for t3.small easily. No sense in maintaining the stress test if it's too wild to be useful.

I understand the distinction between `load` and `stress` test here where `load` is for sustained normal traffic, and `stress` is the type of testing designed to find the limits of the system. If the load test is already behaving like the stress test, another option is to redefine the parameters of both the load and stress tests. In practice that means lowering the amount of concurrent requests. However if we are at a point were we can't really sustain any load, that is another story and it would make a lot of sense to remove the stress test (or would it be the load test that gets removed because in this situation the load test is behaving like a stress test and should be renamed?).

I can see why keeping the stress test could be helpful in the future (What if Saving Satoshi becomes a WORLDWIDE PHENOMENON?!!?! What if we have to drastically scale the instance size up in addition to the autoscaling and the distinction between load and stress tests is greater?!?!!) however as they say "premature optimization is the root of all evil" so I support removing the stress test profile on the grounds that it's code we don't need to burden ourselves with maintaining.

> If you or anyone else has feedback on the testing "scenarios" or traffic patterns, it would be nice to have them match reality.

+1 to what Jamal said about the majority of users doing the same thing at the same time. We've only done 1 Saving Satoshi workshop that was self directed and participants went off and played on their own. That was the 50 person one at TABConf in October 2024. From a system stability perspective that was an absolute disaster, but I think at the time it had more do with actual issues in how the REPL code was architected and less to do with the instance size. Since then [the workshops have all](https://bitcoindevs.gitbook.io/saving-satoshi-in-person-workshop-guide/past-workshops) been led by a facilitator that walks the group through the chapters. This was intentional as participants would be less likely to actively follow along on their computer and more inclined to just passively watch. Also worth noting is when we do workshops, we only cover chapters 1-3. There is only 1 significant REPL activity and that is in chapter 2 where the player needs to write a loop to increment a nonce and compute a hash (aka mining).

For the 2026 BOSS program we made everyone play Saving Satoshi. Between December 2 and January 12, 2000 people played the game and 600 finished it. The experience resulted in MANY issues on both the front and backend coming to the surface (which is how we got to working with you and jamaljsr! :smile_cat: ). 

The only analytics we have for Saving Satoshi [can be found here](https://visits.bitcoindevs.xyz/share/zFmD5WIus09mDxEf/Saving%20Satoshi).

Here's a screenshot of the activity during the phase of the BOSS program when people were playing it:
Screenshot from 2026-05-04 00-18-38

Some rough back-of-the-napkin math says at peak we had 400+ daily players. Divide that by 24 hours in a day and at minimum we had at least 15 people playing concurrently each hour. The real number is likely much higher as we had large groups of players from the same regions of the world sharing timezones. @0tuedon scaled the instance up numerous times, and went far beyond t3.small. I can't recall exactly where we landed but I can either check chat history or our AWS billing history to find out, just let me know if that information would be useful and I will find it. 

Getting back to the question about traffic patterns though, in this scenario, people were playing all different parts of the game at the same time, unlike the TABConf workshop where many of them were working on a specific exercise simultaneously. Also worth noting is that the REPL exercises are the most difficult ones, so if we are going to have concurrent users hammering the system, there is a high likelihood that they will be stuck on a REPL exercise that involves something like the ECC multiplication that you generated for the sample REPL code. We haven't done an analysis of the most "popular" exercises, and by popular I mean most challenging and difficult to pass, but can definitely dig deeper into that if the information is useful to you.

One thing I want to mention is Chapter 8 involves loading a file on the backend and [this was causing a lot of timeouts](https://github.com/saving-satoshi/saving-satoshi/issues/1295). If we wanted to replicate one more scenario, this might be worthwhile. Not sure if that would be in the scope of this PR, and I'm happy to make a ticket for it for myself or someone else to pick up later. With this in mind, I do believe overall the ECC multiplication you have generated for the sample REPL code is an appropriate reflection of the kind scenario that the websocket/REPL part of the system would be under with concurrent users.

**Review by satsie — APPROVED** [review]
Ran the smoke, load test, and repl test locally on Ubuntu 24.04. Did not run the stress test because it's all the same test scenarios, just with a different number of reqs/sec and SLAs and the stress test is designed to push our AWS instance to the limit... not my computer, hah!

This is a fabulous PR and I'm happy to see it merged whenever you feel comfortable. Many thanks to @jrakibi for helping iron out some of those earlier issues with regard to running these performance tests. The tests ran seamlessly for me on first try and I have you both to thank for that. So excited to finally have visibility into how much load the system can handle. Thank you @s373nZ. We are very lucky to have your contributions!

I'll probably wait to see if we can schedule a call for the autoscaling stuff before I dive too deeply into it. It's cool how these tests are able to create a scenario where the autoscaling can be tested.

**Review by jrakibi — APPROVED** [review]
ACK cb4ceec60f719a93a39d677e4fe063f43ca78e46

though I still couldn't run the tests on Mac after the switch to `127.0.0.1`, they only worked once I changed the target to `host.docker.internal`. I think that's fine, macOS users can just override the variables on the command line instead

**Comment by s373nZ** [comment]
* Toned down the HTTP load test to give a slightly more forgiving initial ramp-up period
* Removed the HTTP stress test
* Reviewed HTTP documentation comments after tweaks for accuracy
* Added validation to the REPL distribution knob variables to ensure they total to 1

File: test/performance/processors/functions.js (hunk 5 above)
Comment (s373nZ) [reply test/performance/processors/functions.js:53]: Good question. If they totaled < 1, the distribution is biased to more timeouts (not good!). If they totaled to > 1, the timeouts are eliminated. In both cases, the behavior change is silent to the user. Added validation to ensure the total is one and the user's choice is explicit.

File: test/performance/processors/functions.js (hunk 7 above)
Comment (s373nZ) [reply test/performance/processors/functions.js:101]: This decision is also a bit aggressive, as running elliptic curve multiplication takes 100% of a vCPU and this won't be the only case for all REPL activity (may actually be an outlier), but it feels like it's important to model real activity.

File: test/performance/profiles/repl-capacity.yaml (hunk 8 above)
Comment (s373nZ) [reply test/performance/profiles/repl-capacity.yaml:54]: I just want to confirm that against *these* tests, we are absolutely CPU bound due to the REPL processes. Currently production is on a `t2.small`, but I've been testing against a `t3.small` with double the CPU and half the memory at the same price.

Also, I've been testing against one of those free Supabase hosted databases, and after some time, when more instances join during high traffic, the database starts to become an issue, so that's something to think about too.

File: test/performance/profiles/repl-capacity.yaml (hunk 9 above)
Comment (s373nZ) [reply test/performance/profiles/repl-capacity.yaml:159]: We also test this more reliably in an integration test [here](https://github.com/saving-satoshi/saving-satoshi-backend/blob/19a8cf1b83185f6c3a0baa44d3947ba8cab81dd9/test/integration/websocket/repl.test.ts#L124).

**Comment by s373nZ** [comment]
Cleaned up the documentation some more. I'm feeling pretty good about this and will plan to merge it tomorrow.

@satsie your descriptions of the workshop experiences and BOSS activity are very helpful. I'd say that the huge advantage the Saving Satoshi app has is the ability to plan for these events in advance. Prior to workshop days, or the beginning of the BOSS program, I would just upgrade the instance size variable to a `c6a.xlarge` instance to start with. It would cost about $3.67 / day and features double the CPU power which are dedicated. Meaning there is no performance variable around shared / burstable dynamics.

We can discuss alternative options after we get a chance to demo the autoscaling, I hope.

Thanks to you both, @satsie and @jrakibi for your help and attention!
